[extension/ecsobserver] Add service and task definition based matcher

pingleig opened the pull request:
**Description:**

- Implement match task by service name (like k8s deployment name)
- Implement match task by task definition (btw: it is not same as service name, like pod template 'name')

**Link to tracking Issue:**

- Feature request #1395 

Pending PRs

- Exporter https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3333

Previous PRs

- Fetch Task https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3284
- Docker Label Matcher https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3276
- Task IP and Port https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3133
- Target and Task definition (merged) https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2939
- Config PR (merged) https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2377
- Original PR (closed) #2734 

**Testing:** 

unit test

**Documentation:** 

No new doc. See [existing doc](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/observer/ecsobserver/README.md)

pingleig committed:
ext: ecsobserver Add Service Matcher

pingleig committed:
ext: ecsobserver Add TaskDefinition Matcher

pingleig committed:
ext: ecsobserver Add init all matcher logic

Will be used in sd.go when fetcher is ready

pingleig commented:
cc @Aneurysm9 @anuraaga

anuraaga commented on extension/observer/ecsobserver/service.go:
```diff
@@ -14,12 +14,86 @@
 	// Otherwise both service and container name petterns need to metch.
 	ContainerNamePattern string `mapstructure:"container_name_pattern" yaml:"container_name_pattern"`
+
+	// should never be nil because Init must reject it and caller should stop
```
Not sure how to interpret "should" here

pingleig replied:
Moved the regex into matcher impl.

anuraaga commented on extension/observer/ecsobserver/service.go:
```diff
@@ -14,12 +14,86 @@
+	containerNameRegex *regexp.Regexp
+}
+
+func (s *ServiceConfig) Init() error {
```
Do these need to be public?

anuraaga replied:
Actually this seems a bit weird since we're reusing the config to store non-config fields, I don't think we generally do that. Can you use a different private struct for the initialized config?

pingleig replied:
Renamed it to Validate, and put the initialized value into matcher impl.

anuraaga commented on extension/observer/ecsobserver/service.go:
```diff
@@ -14,12 +14,86 @@
+	if t.Service == nil {
+		return nil, errNotMatched
+	}
+	// nameRegex should never be nil
```
This repeated comment is a code smell, I guess my suggestion above of using two types so you guarantee a fully initialized type will help

pingleig replied:
Done, see comment above (moved to the matcher struct)

pingleig committed:
ext: ecsobserver Move internal states into Matcher impl

- Implement `Validate` on top level config struct so error should show
  up earlier when loading the config.

pingleig commented on extension/observer/ecsobserver/matcher.go:
```diff
@@ -135,3 +176,26 @@ func matchContainers(tasks []*Task, matcher Matcher, matcherIndex int) (*MatchRe
+					MetricsPath: expCfg.MetricsPath,
+					Job:         expCfg.JobName,
+				})
+			}
```
Add a map in the initialized common export struct to avoid the nested loop, the map is created when checking duplicated metrics port.

pingleig commented on extension/observer/ecsobserver/matcher.go:
```diff
@@ -88,6 +105,30 @@ type MatchedTarget struct {
 }
 
+func newMatchers(c Config, mOpt MatcherOptions) (map[MatcherType][]Matcher, error) {
+	// We can have a registry or factory methods etc. but since we only have three type of metchers in filter.
```
Added I decided to hard code the map here because the three matchers will likely be all the matchers.

pingleig commented on extension/observer/ecsobserver/matcher.go:
```diff
@@ -88,6 +105,30 @@ type MatchedTarget struct {
+	}
+	matchers := make(map[MatcherType][]Matcher)
+	matcherCount := 0
+	for tpe, cfgs := range matcherConfigs {
```
go's parser does not allow using `type` as variable name will get `syntax error: unexpected type, expecting expression`

https://play.golang.org/p/dygnGWUgKNR

anuraaga replied:
In that case `t` is still better, `tpe` looks like a three letter acronym

pingleig replied:
Changed to `mType`

anuraaga commented on extension/observer/ecsobserver/config.go:
```diff
@@ -52,6 +53,31 @@ type Config struct {
 }
 
+// Validate overrides the embedded noop validation.
+func (c *Config) Validate() error {
```
Make sure any new method added that are only called in this package are private, I think `Validate` is at least. Should do a scrub of existing methods / structs soon too but not for this PR, we should take care of new ones though.

pingleig replied:
This one has to be `Validate` so it will get called when loading config automatically  https://github.com/open-telemetry/opentelemetry-collector/blob/0bbb1b7f3d886c4ba7a03b36ce4b57c538a727c5/config/extension.go#L17-L24

pingleig replied:
I made all the internal validation logic private, so the sub config for matchers are exported but all the methods `validate` and `newMatcher` are private now.

pingleig committed:
ext: ecsobserver Make matcherConfig internal interface

pingleig commented:
ping @anuraaga btw: there is also #3333
